osquery/config/packs.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ *  Copyright (c) 2014, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/property_tree/json_parser.hpp>
+
+#include <osquery/core.h>
+#include <osquery/config/packs.h>
+#include <osquery/logger.h>
+#include <osquery/sql.h>
+
+
+namespace osquery {
+
+Pack::Pack(const std::string& name, const ptree& tree) {
+  initialize(name, "", tree);
+}
+
+Pack::Pack(const std::string& name,
+           const std::string& source,
+           const ptree& tree) {
+  initialize(name, source, tree);
+}
+
+Pack::Pack(const std::string& name, const std::string& json) {
+  std::stringstream stream;
+  stream << json;
+  ptree tree;
+  pt::read_json(stream, tree);
+  initialize(name, "", tree);
+}
+
+Pack::Pack(const std::string& name,
+           const std::string& source,
+           const std::string& json) {
+  std::stringstream stream;
+  stream << json;
+  ptree tree;
+  pt::read_json(stream, tree);
+  initialize(name, source, tree);
+}
+
+void Pack::initialize(const std::string& name,
+                      const std::string& source,
+                      const ptree& tree) {
+  name_ = name;
+  source_ = source;
+  discovery_queries_.clear();
+  if (tree.count("discovery") > 0) {
+    for (const auto& item : tree.get_child("discovery")) {
+      discovery_queries_.push_back(item.second.get_value<std::string>());
+    }
+  }
+
+  platform_.clear();
+  if (tree.count("platform") > 0) {
+    platform_ = tree.get<std::string>("platform");
```

Comment:
Caution! This will throw if `"platform"` cannot be casted to a string. You can avoid the `count` and `get<std::string>` and account for the failed cast using:

```cpp
platform_ = tree.get("platform", "");
```

The type of the second parameter is the inferred template type. Saves you a clear too!